app/scripts/lib/AbstractPetnamesBridge.ts

```diff
@@ -0,0 +1,240 @@
+  #isTwoWay: boolean;
+
+  #nameController: NameController;
+
+  #synchronizingDirection: SyncDirection | 'NONE' = 'NONE';
+
+  #messenger: PetnamesBridgeMessenger;
+
+  /**
+   * @param options
+   * @param options.isTwoWay - Indicates whether the bridge is two-way or not. One-way bridges are Source->Petnames only.
+   * @param options.nameController
+   * @param options.messenger
+   */
+  constructor({
+    isTwoWay,
+    nameController,
+    messenger,
+  }: {
+    isTwoWay: boolean;
+    nameController: NameController;
+    messenger: PetnamesBridgeMessenger;
+  }) {
+    this.#isTwoWay = isTwoWay;
+    this.#nameController = nameController;
+    this.#messenger = messenger;
+  }
+
+  // Initializes listeners
+  init(): void {
+    if (this.#isTwoWay) {
+      this.#messenger.subscribe('NameController:stateChange', () =>
+        this.#synchronize('Petnames->Source'),
+      );
+    }
+    this.onSourceChange(() => this.#synchronize('Source->Petnames'));
+  }
+
+  /**
+   * Adds a listener for source change events.
+   *
+   * @param listener - The listener function to be called when a source change event occurs.
+   */
+  protected abstract onSourceChange(listener: () => void): void;
+
+  /**
+   * Retrieves the source entries.
+   *
+   * @returns An array of PetnameEntry objects representing the source entries.
+   */
+  protected abstract getSourceEntries(): PetnameEntry[];
+
+  /**
+   * Update the Source with the given entry. To be overridden by subclasses.
+   *
+   * @param type
+   * @param entry
+   */
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  protected updateSourceEntry(type: ChangeType, entry: PetnameEntry): void {
+    throw new Error('updateSourceEntry must be overriden for two-way bridges');
+  }
+
+  /**
+   * Synchronizes Petnames with the Source or vice versa, depending on the direction.
+   *
+   * @param direction - The direction to synchronize in.
+   */
+  #synchronize(direction: SyncDirection): void {
+    if (this.#synchronizingDirection === direction) {
+      throw new Error(
+        `synchronize(${direction}) called recursively in same direction`,
+      );
+    }
+    if (this.#synchronizingDirection !== 'NONE') {
+      return; // Ignore calls while updating in the opposite direction
+    }
+
+    this.#synchronizingDirection = direction;
+
+    const [newEntries, prevEntries] =
+      direction === 'Source->Petnames'
+        ? [this.getSourceEntries(), this.#getPetnameEntries()]
+        : [this.#getPetnameEntries(), this.getSourceEntries()];
+
+    const changeList = this.#computeChangeList(prevEntries, newEntries);
+    this.#applyChangeList(changeList);
+
+    this.#synchronizingDirection = 'NONE';
+  }
+
+  /**
+   * Extract PetnameEntry objects from the name controller state.
+   */
+  #getPetnameEntries(): PetnameEntry[] {
+    const { names } = this.#nameController.state;
+    const entries: PetnameEntry[] = [];
+    for (const type of Object.values(NameType)) {
```

Comment:
These bridges are all specific to `NameType.ETHEREUM_ADDRESS` so we don't need to query the names for any other name types.

Arguably we could call this `AbstractAddressPetnamesBridge` but that feels like overkill as it's the only current `NameType`.